Allow WMR to UC cross-workspace copying of model versions (#17458)
mlflow/mlflow: docs/docs/classic-ml/model-registry/workflow.mdx

mlflow.set_registry_uri("databricks-uc://my-databricks-shard1")
```

#### Migrating to Databricks Unity Catalog Model Registry from Databricks Workspace Model Registry

To migrate model versions from the Databricks Workspace Model Registry to the Databricks Unity Catalog Model Registry, we
recommend using the <APILink fn="mlflow.client.MlflowClient.copy_model_version">`copy_model_version()`</APILink> API. The
source URI would be from your workspace registry, and the destination would be a 3-level UC model name, e.g.

```python
from mlflow import MlflowClient

# Registry URI must be set to workspace registry
client = MlflowClient(registry_uri="databricks")
src_model_uri = f"models:/my_wmr_model/1"
uc_migrated_copy = client.copy_model_version(
    src_model_uri, "mycatalog.myschema.my_uc_model"
)
```

> If the destination UC model doesn't exist, one will be created as part of this API call (assuming the right ACLs).

Reviewer: ```suggestion
If the destination UC model doesn't exist, one will be created as part of this API call.
```

Additionally, models registered to Databricks Unity Catalog require a signature to be set. If the workspace-registered
model version does not already have a signature set, we recommend following the [Model Signatures](/ml/model/signatures)

> instructions to add one. However, to ease the burden of migration, you can also set the environment variable

Reviewer: ```suggestion
instructions to add one. However, to simplify the migration process, you can also set the environment variable
```

> `MLFLOW_SKIP_SIGNATURE_CHECK_FOR_MIGRATION_TO_DATABRICKS_UC_REGISTRY` to `true` that will work ONLY with the

Reviewer: ```suggestion
`MLFLOW_SKIP_SIGNATURE_CHECK_FOR_MIGRATION_TO_DATABRICKS_UC_REGISTRY` to `true`. This ONLY works with the
```

<APILink fn="mlflow.client.MlflowClient.copy_model_version">`copy_model_version()`</APILink> API as indicated above, and
allow you to migrate the model to Unity Catalog, bypassing the signature requirement.

> NOTE: not having a signature on your Unity Catalog registered model version will limit the available feature set.

Author: Synced offline, will remove for now and then re-add after docs are published

```python
import os

skip_sig_var = "MLFLOW_SKIP_SIGNATURE_CHECK_FOR_MIGRATION_TO_DATABRICKS_UC_REGISTRY"

> os.environ[skip_sig_var] = "true"

Author: Yup that was exactly the issue, added a lint check exemption

> os.environ[skip_sig_var] = "true"

Reviewer 2: can we rename this environment variable to `MLFLOW_VALIDATE_SIGNATURE_FOR_MIGRATION_TO_DATABRICKS_UC_REGISTRY`? Does that fit in a single line?

> os.environ[skip_sig_var] = "true"

Author: Yeah good name! Renamed it to that variable and looks like it works

```

If you want to migrate all of the model versions in your workspace registered model to a destination UC model,
you can use a script as follows:

```python
import mlflow
from mlflow import MlflowClient
from mlflow.models import ModelSignature
from mlflow.types.schema import Schema, ColSpec, AnyType

workspace_client = MlflowClient(registry_uri="databricks")
uc_client = MlflowClient(registry_uri="databricks-uc")

# Make a placeholder model that can be used to increment the version number
def make_placeholder_uri() -> str:
    class _Placeholder(mlflow.pyfunc.PythonModel):
        def predict(self, ctx, x):
            return None

    with mlflow.start_run() as run:
        schema = Schema([ColSpec(AnyType())])
        model = mlflow.pyfunc.log_model(
            name="m",
            python_model=_Placeholder(),
            signature=ModelSignature(inputs=schema, outputs=schema),
        )
        return f"models:/{model.model_id}"

# Check if the source model has a particular version number

> def src_has_mv(name: str, v: int) -> bool:

Reviewer: ```suggestion
def workspace_model_exists(name: str, version: int) -> bool:
```

    try:
        workspace_client.get_model_version(name, str(v))
        return True

>     except Exception:

Author: Updated to catch only that exception and error code, and manually verified it works!

        return False

# Copy model versions from a source Databricks workspace-registered model to
# a destination Databricks Unity Catalog registered model
def copy_model_versions_to_uc(src: str, dst: str) -> None:

>     latests = workspace_client.get_latest_versions(src)

Reviewer: ```suggestion
    latest_versions = workspace_client.get_latest_versions(src)
```

    max_version_number = max(int(v.version) for v in latests)

>     placeholder_uri = make_placeholder_uri()

Reviewer: ```suggestion
    placeholder_model = make_placeholder_model()
```

    for v in range(1, max_version_number + 1):
        if src_has_mv(src, v):
            workspace_client.copy_model_version(f"models:/{src}/{str(v)}", dst)
        else:
            # Create and immediately delete a placeholder model version to increment
            # the version counter on the UC model, so the version numbers on the UC
            # model match those on the workspace registered model.
            mv = uc_client.create_model_version(dst, placeholder_uri)
            uc_client.delete_model_version(dst, mv.version)

> copy_model_versions_to_uc("my_wmr_model", "mycatalog.myschema.my_uc_model")

Reviewer: ```suggestion
copy_model_versions_to_uc("my_workspace_model", "mycatalog.myschema.my_uc_model")
```

<APILink fn="mlflow.client.MlflowClient.copy_model_version">`copy_model_version()`</APILink> API as indicated above, and
allow you to migrate the model to Unity Catalog, bypassing the signature requirement.

```python
import os

skip_sig_var = "MLFLOW_SKIP_SIGNATURE_CHECK_FOR_MIGRATION_TO_DATABRICKS_UC_REGISTRY"
os.environ[skip_sig_var] = "true"
```

NOTE: not having a signature on your Unity Catalog registered model version will limit the available feature set.
For a full list of what will be unavailable, refer to the [Databricks documentation](https://docs.databricks.com/aws/en/machine-learning/manage-model-lifecycle/#train-and-register-unity-catalog-compatible-models).

If you want to migrate all of the model versions in your workspace registered model to a destination UC model,
you can use a script as follows:

```python
import mlflow
from mlflow import MlflowClient
from mlflow.models import ModelSignature
from mlflow.types.schema import Schema, ColSpec, AnyType

workspace_client = MlflowClient(registry_uri="databricks")
uc_client = MlflowClient(registry_uri="databricks-uc")

# Make a dummy model that can be used to increment the version number

> def make_dummy_uri() -> str:

Reviewer: Can we call this `placeholder` instead of dummy?

    class _Dummy(mlflow.pyfunc.PythonModel):
        def predict(self, ctx, x):
            return None

    with mlflow.start_run() as run:
        schema = Schema([ColSpec(AnyType())])
        model = mlflow.pyfunc.log_model(
            name="m",
            python_model=_Dummy(),
            signature=ModelSignature(inputs=schema, outputs=schema),
        )
        return f"models:/{model.model_id}"

# Check if the source model has a particular version number
def src_has_mv(name: str, v: int) -> bool:
    try:
        workspace_client.get_model_version(name, str(v))
        return True
    except Exception:
        return False

# Copy model versions from a source Databricks workspace-registered model to
# a destination Databricks Unity Catalog registered model
def copy_model_versions_to_uc(src: str, dst: str) -> None:
    latests = workspace_client.get_latest_versions(src)

>     top = max(int(v.version) for v in latests)

Reviewer: ```suggestion
    max_version_number = max(int(v.version) for v in latests)
```

    dummy_uri = make_dummy_uri()

    for v in range(1, top + 1):
        if src_has_mv(src, v):
            mlflow.set_registry_uri("databricks")
            workspace_client.copy_model_version(f"models:/{src}/{str(v)}", dst)
        else:

>             mlflow.set_registry_uri("databricks-uc")

Reviewer: We don't need to call `set_registry_uri()` if we're using `MlflowClient` instances

Author: Yeah now that we updated to using the store's URI, we don't need to any more

            mv = uc_client.create_model_version(dst, dummy_uri)
            # delete immediately so it just bumps numbering

>             uc_client.delete_model_version(dst, mv.version)

Reviewer: Can we add a more detailed inline comment to explain what's going on here?

NOTE: not having a signature on your Unity Catalog registered model version will limit the available feature set.
For a full list of what will be unavailable, refer to the [Databricks documentation](https://docs.databricks.com/aws/en/machine-learning/manage-model-lifecycle/#train-and-register-unity-catalog-compatible-models).

If you want to migrate all of the model versions in your workspace registered model to a destination UC model,
you can use a script as follows:

```python
import mlflow
from mlflow import MlflowClient
from mlflow.models import ModelSignature
from mlflow.types.schema import Schema, ColSpec, AnyType

workspace_client = MlflowClient(registry_uri="databricks")
uc_client = MlflowClient(registry_uri="databricks-uc")

# Make a dummy model that can be used to increment the version number
def make_dummy_uri() -> str:
    class _Dummy(mlflow.pyfunc.PythonModel):
        def predict(self, ctx, x):
            return None

    with mlflow.start_run() as run:
        schema = Schema([ColSpec(AnyType())])
        mlflow.pyfunc.log_model(
            name="m",
            python_model=_Dummy(),
            signature=ModelSignature(inputs=schema, outputs=schema),
        )

>         return f"runs:/{run.info.run_id}/m", run.info.run_id

Author: Yeah good point, updated it

mlflow.set_registry_uri("databricks-uc://my-databricks-shard1")
```

#### Migrating to Databricks Unity Catalog Model Registry from Databricks Workspace Model Registry

To migrate model versions from the Databricks Workspace Model Registry to the Databricks Unity Catalog Model Registry, we
recommend using the <APILink fn="mlflow.client.MlflowClient.copy_model_version">`copy_model_version()`</APILink> API. The
source URI would be from your workspace registry, and the destination would be a 3-level UC model name, e.g.

```python
from mlflow import MlflowClient

# Registry URI must be set to workspace registry
client = MlflowClient(registry_uri="databricks")
src_model_uri = f"models:/my_wmr_model/1"
uc_migrated_copy = client.copy_model_version(

>     src_model_uri, "mycatalog.myschema.my_uc_model"

Reviewer: Does the destination model need to be created first? If so, can we clarify and show users how to create it?

Author: It does not actually, we'd create it if it doesn't exist already

Reviewer: Got it; can we mention that in the example?

> )

Author: Added

```

Additionally, models registered to Databricks Unity Catalog require a signature to be set. If the workspace-registered
model version does not already have a signature set, we recommend following the [Model Signatures](/ml/model/signatures)
instructions to add one. However, to ease the burden of migration, you can also set the environment variable
`MLFLOW_REGISTRY_MIGRATION_SKIP_SIGNATURE_VALIDATION` to `true` that will work ONLY with the

<APILink fn="mlflow.client.MlflowClient.copy_model_version">`copy_model_version()`</APILink> API as indicated above, and
allow you to migrate the model to Unity Catalog, bypassing the signature requirement.

```python
import os

> os.environ["MLFLOW_REGISTRY_MIGRATION_SKIP_SIGNATURE_VALIDATION"] = "True"

Reviewer: Can we support lowercase `"true"` too?

```

NOTE: not having a signature on your Unity Catalog registered model version will limit the available feature set.

> For a full list of what will be unavailable, refer to the [Databricks documentation](https://docs.databricks.com/aws/en/machine-learning/manage-model-lifecycle/#train-and-register-unity-catalog-compatible-models).

Author: Working with @andreakress to update these docs, placeholder link that talks about signatures today in the meantime

Reviewer: Right - the linked docs don't say anything about which features are unavailable. What features are unavailable? If we can't point to anything really significant, let's remove this. We can add it back later if / when needed.

Author: Kk FYI @andreakress, we'll add this back in after we publish the unavailable feature set on the Databricks docs
